accord-core/src/main/java/accord/messages/Apply.java

```diff
@@ -54,15 +63,59 @@ public Apply(Node.Id to, Topologies topologies, TxnId txnId, Txn txn, Key homeKe
         this.result = result;
     }
 
+    @VisibleForImplementation
+    public Apply(Keys scope, long waitForEpoch, TxnId txnId, Txn txn, Key homeKey, Timestamp executeAt, Deps deps, Writes writes, Result result)
+    {
+        super(scope, waitForEpoch);
+        this.txnId = txnId;
+        this.txn = txn;
+        this.homeKey = homeKey;
+        this.executeAt = executeAt;
+        this.deps = deps;
+        this.writes = writes;
+        this.result = result;
+    }
+
+    static Future<Void> waitAndReduce(Future<Void> left, Future<Void> right)
```

Comment:
Reply from Blake in JIRA

Good point. Changed return type to Future<Void>

Reply:
Are we sure this is the best way to do this? We're occupying a thread waiting for these results. My preference would be to supply a callback to the apply method, as this is also easier for me to refactor to enrich with extra parameters explaining the outcome in my partial replication patch, but alternatively listeners would suffice?

If I don't hear anything I will go ahead with refactoring to callbacks in my follow-up work.

Reply:
In fact, in this particular case, do we even need the execute outcome? If the apply is successful we can respond immediately to the peer?

Reply:
Good point about blocking the request thread, I'll fix that. Regarding callbacks vs futures (here and returned from Read#read), I usually prefer futures since returning them is enforced by the compiler, whereas forgetting to call a callback may not be obvious until runtime. Command#maybeExecute is a good example of a method where it’s easier to make mistakes with callbacks that the compiler won’t let you make with futures. I also find it a bit easier to reason about which method needs to worry about completing a future vs callbacks. Why do you prefer callbacks for the read and write methods? Is it because you can require an arbitrary number of arguments to the callback?

Also, in this case we don't need the execute outcome (and it's Void) we just need to know when we're done so we can respond to the coordinator.

Reply:
I don’t see how Futures provide any additional compiler enforcement? The future here is simply a carrier for a callback, since we must not block the thread - if anything we have demonstrated a very easy way we can misuse the Future API, that isn’t possible with a callback. The callback can be required as a parameter, so it is enforced to be provided the same as a Future must be returned. The Future itself must be completed with a parameter and this is just as weakly enforced (in fact identical in requirements) to invoking a callback with the parameters, so I don’t understand the perceived benefit here.

A Future is however:
1. Harder to debug, especially when chaining several together as the callstack becomes very polluted
2. Much less efficient (in both allocations and in cost to maintain and invoke)
3. Much less flexible - as mentioned, I cannot readily add any additional parameters. We have to introduce an additional layer of indirection to carry additional information - and additional allocations
4. Most of the time - including in this instance - just serves as an additional layer of indirection to implement a callback.

I don’t think Future should never be used, but I think it’s use should be minimised to cases where it’s really needed, or to APIs where we need to provide greater flexibility to consumers.

Reply:
The compiler helps in the `maybeExecute` scenario I was talking about, where there are cases where the method bails out before the main event. For example:

```java
// it's easy to forget to invoke the callback here
if (shouldBailOut)
{
    callback.accept(null, null);
    return;
}

// but you have to return something here
if (shouldBailOut)
    return ImmediateFuture.success(null);
```

Of course, you have to remember to invoke the callback on the "main event" of execution as you do the future, although oftentimes things executor provided wrappers will do that for you, but I do believe the return signature does help prevent mistakes like the one above. 

Also, to be fair, it's possible to abuse the a callback in the same way as it was abused in this instance, by blocking the thread while waiting for it to be invoked.

In the case of the apply method here, in C* this terminates in a call submitting a mutation to the mutations stage, which returns a future. This future is then cached so we can de-dupe multiple apply requests for a given command (same on the read side). This _could_ be achieved by manually managing callbacks, but it will be more complicated and will end up reimplementing a lot of the future callback logic.

> Harder to debug, especially when chaining several together as the callstack becomes very polluted

The callstack does get deeper, but I can usually skip over the future specific stuff without too much trouble

> Much less efficient (in both allocations and in cost to maintain and invoke)

Maybe in raw opcode/byte count and/or some microbenchmarks... but in the context of the larger system, I don’t think ditching futures for callbacks would be a big performance win.
> Much less flexible - as mentioned, I cannot readily add any additional parameters. We have to introduce an additional layer of indirection to carry additional information - and additional allocations

How many extra parameters are we talking about here? The future isn’t returning a value as is, so you get one for free.

> Most of the time - including in this instance - just serves as an additional layer of indirection to implement a callback.

I wouldn’t call them “just an additional layer of indirection”, they’re tools that simplify programmer management of callbacks and their edge cases, albeit at the cost of another layer of indirection.

Reply:
> The callstack does get deeper, but I can usually skip over the future specific stuff without too much trouble

Unfortunately, I can't. I find it very challenging to debug nested `Future` callbacks, and I think this should by itself be a strong reason to avoid them except where they provide significant ergonomic advantages. But, if this were the only consideration I would perhaps accept it on taste grounds. Most of the other points fall similarly... But

> Maybe in raw opcode/byte count and/or some microbenchmarks... but in the context of the larger system, I don’t think ditching futures for callbacks would be a big performance win.

This is the main thing I have to disagree with strongly. This kind approach really hamstrings Cassandra's ability to improve. Each Future with callback is an additional three CAS instructions and ~100 bytes of memory. This is a hot path, where we already removed a chain of about four `Future` when internally processing requests. This is the removal of perhaps a dozen CAS instructions and several hundred bytes allocated. This is unlikely to be very measurable _today_, but only because of hundreds of accumulated decisions that deemphasised this kind of thing.

I am in favour of this kind of inefficiency when there's _strong justification_. But when it's at best a toss-up (in my opinion, it is worse), and you can convert a callback to a `Future` _for free_ but can never save the `Future` cost once the API has committed you to it, then I feel strongly that this isn't the right approach.

>In the case of the apply method here, in C* this terminates in a call submitting a mutation to the mutations stage, which returns a future.

This is true today, but it feels like a bit of a dog wagging the tail situation. I would prefer not to base the API on the current implementation. It is not clear this is how we _will_ be doing things (esp. with improved commit log likely preceding query processing, so that we only have to touch memtables).

Reply:
> This kind approach really hamstrings Cassandra's ability to improve. 

You mean a detailed and informed discussion of the pros and cons of different solutions? I feel like the project could use more of this ;)

> Each Future with callback is an additional three CAS instructions and ~100 bytes of memory. This is a hot path, where we already removed a chain of about four Future when internally processing requests. This is the removal of perhaps a dozen CAS instructions and several hundred bytes allocated. 

I understand this is a hot path, and prior to pushing back on this, I implemented a callback class that would meet the needs of apply, and the cas/bytes numbers above are way overstated. The way we distribute work to command stores negates a lot of the potential simplicity of callbacks. Specifically, we may have to talk to multiple command stores, and we don’t know how many until we’ve dispatched our tasks.

In the case where we interact with a single command store, futures actually use slightly less memory than callbacks. In the worst case, a future will use 24 more bytes and one more cas operation than a callback. 

Here’s the callback class if you’d like to take a look, it’s possible I’ve overlooked something: https://gist.github.pie.apple.com/beggleston/56e540bdb6408dd7f22ec88d095c42e2

In the more complex use of futures, like this, callback state starts to look a lot like future state.

Reply:
To articulate my position here in some more detail, I believe we should avoid practices that are hazardous, or that make it easier for bugs to be introduced in the future. From this perspective, I see callbacks as inferior to futures because they make control flow the responsibility of the programmer in a way that is less obvious, easier to forget about, and more difficult to reason about than futures. 

While futures also make control flow the responsibility of the programmer, they’re typically on a much shorter leash. Often being instantiated and completed in a much narrower scope, like an executor, or in some critical path, where the focus very clearly includes correctly handling control flow via the future.

In the case of apply, we’re talking about instantiating a callback in the message process method, plumbing it through a decent chunk of the command code, and into the storage layer. In my view, this is an unacceptably large surface area to be requiring programmers to be handling an opaque callback correctly.

I agree that the project has historically had a problem with not taking performance seriously, but I think it has had an even bigger problem with not taking safely seriously, and has also suffered more because of it.

I think it’s fine to trade safety for efficiency when the risk is low and the gains are non-trivial, but I feel pretty strongly that that isn’t the case here.

Reply:
> You mean a detailed and informed discussion of the pros and cons of different solutions? I feel like the project could use more of this ;)

No, absolutely :)

> In the case where we interact with a single command store, futures actually use slightly less memory than callbacks.

That definitely means we've done callbacks wrong :) ... bear in mind, the `Future` has to allocate a list entry for the callback, potentially allocates a `WaitQueue` for consumers like those above, and it has to perform a CAS to add the callbacks, signal the callbacks, and to set the result (also at least four for each waiting `Thread`). At _most_ a callback should have to do the last of these (and usually not even that).

> Here’s the callback class if you’d like to take a look

This looks like it is trying to implement a map/Reduce chain over a list of callbacks, i.e. the `waitAndReduce` method - rather than a single callback? But we have to solve this anyway still above, as we shouldn't be blocking, I think. So we will need some kind of mechanism for this either way, don't we? I think it can be a lot simpler though.

> Often being instantiated and completed in a much narrower scope, like an executor, or in some critical path

So, literally anywhere you can have a `Future` you can have a callback with exactly the same control flow. There's no need for extra plumbing. Wherever you would invoke `AbstractFuture.setSuccess` or `AbstractFuture.setFailure` you can invoke a callback.

It is true that an `Executor` lets you hand this off, but we can do that for callbacks too. We have the power! `ExecutorPlus` can easily be given a callback `execute` method variation, in fact this seems like a really good idea. But also, a lambda can be submitted that simply has a try/catch branch doing the obvious things.

> I'm not saying performance and safety are mutually exclusive, though there's often a trade off between the two.

If I thought `Future` afforded more safety, I'd definitely agree it is a cost worth paying, so in that respect I can absolutely see your point of view. I just still don't see the extra safety here. 

Perhaps we do as a project need to offer as much scaffolding and support for callbacks as we already do for `Future`. Perhaps we can also consider variants of `Future` that are more restricted but efficient (and have easier call stacks!), so we can begin walking the two extremes closer towards each other as well.

Reply:
> That definitely means we've done callbacks wrong :) ... bear in mind, the Future has to allocate a list entry for the callback, potentially allocates a WaitQueue for consumers like those above, and it has to perform a CAS to add the callbacks, signal the callbacks, and to set the result (also at least four for each waiting Thread). At most a callback should have to do the last of these (and usually not even that).

Sure, I know, and I’m not arguing that a simple callback is less efficient than a future. The point I’m trying to make is that you can’t use a simple callback for apply. If you want to use callbacks with apply, avoid locking, have minimal memory allocation, and prevent multiple calls from the same command store from causing problems, you need this extra state and cas operations to do it, and once you’ve added all that, the efficiency argument becomes a lot thinner.

> So, literally anywhere you can have a Future you can have a callback with exactly the same control flow. There's no need for extra plumbing. Wherever you would invoke AbstractFuture.setSuccess or AbstractFuture.setFailure you can invoke a callback.
 
In the places where you’d be calling `setSuccess` or `setFailure`, that’s true. My issue with callbacks applies to everywhere else. Assume that the originator of the future or the destination of the callback will do the right thing, and think of the rest of the program. Since Future encapsulates the future result of an operation you can do one of two things with it, if you need the result, get it, otherwise return it down the stack. 

With a callback, there are more options:
 * If an exception is thrown, do you need to call it? 
 * How do you make sure you don’t call it twice? 
 * What happens if you call it twice? Does it throw an exception, or totally break your program? 

With callbacks, these are now things you need to think about. None of this is intractable, but it creates an unnecessary distraction from whatever consensus thing you’re doing anyway, and with your attention split between consensus and a callback you’re supposed to know what to do with, you’re more likely to make a mistake that you and the reviewer miss.

> Perhaps we can also consider variants of Future that are more restricted but efficient (and have easier call stacks!), so we can begin walking the two extremes closer towards each other as well.

I was going to suggest this as a middle ground, a lot of the inefficiency of our futures classes come from their needing to work properly in all cases

Reply:
>  you need this extra state and cas operations to do it, and once you’ve added all that, the efficiency argument becomes a lot thinner.

Except I think all of this extra work needs to be done for `Future` too? I'm pretty sure the correct approach here is simply using the `Future` as a carrier for a callback? So in this case we have two sides of the safety coin: the side we pass to the `Executor` which can be easily encapsulated safely (or done for us by the `Executor` even more cheaply), and the consumer side which I think will be identical, as we need to manage the concurrency of multiple asynchronous results being accumulated in the object that is reducing them, and we need to do it without blocking the thread?

> With a callback, there are more options:

I don't think there are? The only affordance I see here is the `trySuccess` and `tryFailure` methods, but we can readily have a very lightweight class (a single boolean) that a callback can accept (or be passed to as a lambda) that ensures this isn't a problem.

Obviously there's a huge spectrum of possible functionality/performance points, and we have to be selective so as not to overwhelm our lingua franca. I think we might have room for a couple of additional points in the spectrum though: lightweight `Future` objects designed to only support at most one follow-up action (either a caller waiting on the result, or a listener of some kind); and better callback support.

But I think we should always reach for callbacks first, and only upgrade to `Future` where we find it specifically is well justified by the call-site, so that the trade-offs are well considered and we don't become overly dependent on the additional machinery when it isn't needed (the caller I think can always be made equivalent to `Future`, as discussed above). I think by far the majority of use-cases in Cassandra (esp. until we get virtual threads) are best expressed as single submit/consume transactions with an object accumulating one or more outcomes - and callbacks fit this niche perfectly.

I guess we have accidentally trodden on a minor landmine about project code style and direction :)

Reply:
I guess I should say, though, that my protestations do become a lot more muted with a more efficient `Future`. I do wonder how efficient we can really go, though, having pondered it a little. We can reduce the memory footprint a little, but it is hard to improve the performance due to multi-threading - and anything we do to improve the multi-threading is likely to involve memory consumption trade-offs.

I have a general preference to use the minimal API required to implement a feature, but a really cheap and ergonomic future might be compelling anyway. Perhaps I will have a play with `AbstractFuture` and see what I can extract if we impose some restrictions such as a single follow-up action.

Reply:
> Except I think all of this extra work needs to be done for Future too?

Hmm good point. So that would penalize futures an additional cas instruction when there is more than one future to listen to.

> I don't think there are?

There definitely are. The Future interface(s) do not have trySuccess or tryFailure methods, or any other methods for setting a result, you’re thinking of the implementations. Again, I’m not talking about the places responsible for completing an operation, I’m talking about the rest of the program. Futures only let you receive the outcome of an operation, they’re one sided. They separate the responsibility of performing an operation from the ability to consume its result. The lack of any methods for completing the operation clearly communicate to the programmer that it’s not their responsibility, the originator of the future is responsible for doing the right thing. With callbacks, this separation does not exist.

> But I think we should always reach for callbacks first

I agree, but I think we should upgrade to Futures as soon as the callback needs to be passed around to areas of the program not responsible for completing it.

Reply:
> I think this is the source of our disconnect. Where would this happen? 

In our apply example, a callback would be passed up the entire call stack

```
Apply#process
Command#apply(Txn, Key, Key, Timestamp, Deps, Writes, Result)
Command#maybeExecute
Command#apply(void)
Command#postApply
```

So that’s 3 methods it has to go through before it gets to postApply, which is supposed to call it. 2 of those methods (both Command#apply) are never supposed to call it, but can do so erroneously. Then there’s maybeExecute, which is supposed to call it sometimes. maybeExecute is particularly tricky because there are multiple points where you need to call the callback _and_ return, and if you do one and not the other, you’ve introduced a bug.

Reply:
In this example, I don't think we need to invoke a callback at all in `postApply`. There's no need for the caller to be informed when `Apply` completes - only execute? There might be some error handling to perform in the case that apply later fails, but why are we informing the caller about apply here?

But, leaving that aside, I don't really see the problem? Each method has its own responsibility - if we need to scope it better, we can create a new callback for the relevant scope, that translates the semantics (as we do for `Future`). But I don't see any additional safety - the `Future` we return is just as prone to being done incorrectly at each step? If anything, to me this just highlights additional efficiencies of callbacks, as we _can_ quite safely pass them down without allocating additional objects.

This is exactly what we do for coordination, we pass through the callback until either we hit an exception or we pass responsibility of the callback onwards.

Reply:
For reference, here's a patch introducing a minimal future that is limited to one listener and as many waiters as we like, with zero CAS operations in the normal path, and no unnecessary allocations (but still more weight than a callback)*: https://github.com/belliottsmith/cassandra/tree/minimalfuture

I still think these use cases are better with callbacks though!

*need to double check implementation with a clearer head

Reply:
> But I don't see any additional safety - the Future we return is just as prone to being done incorrectly at each step?

No, you can't erroneously complete a future at each step like you can with a callback.

Reply:
> No, you can't erroneously complete a future at each step like you can with a callback.

This is the safety difference I keep bringing up and I don't feel like it's really being engaged with in this disussion.

The advantages of futures is that they leverage the type system to eliminate a class of bugs that callbacks make possible.

Reply:
I am genuinely struggling to see it. What class of erroneous completions are we avoiding? If maybeExecute does the wrong thing, why would it do the right thing with a Future? Usually we aim to defend against erroneous use by external code, not code we have control over - once we distrust ourselves all bets are off, and compiler assistance here seems very weak.

Is it that you think the returning of an object somehow ensures semantic scope, or that composition methods available to Future somehow make it easier to guarantee correct completion?

If it’s that you may unintentionally complete a callback more than once (eg with exceptional completion), this can be guarded against where we perceive a risk of it with a single cheap encapsulating object, as discussed.

if it’s that you are concerned that a higher method (maybeExecute) may accidentally invoke a callback even after it has passed ownership to a lower method (apply), this can also be cheaply guarded against (by eg nulling out the local parameter when you pass control) - but it’s also not very hard to simply verify it by code inspection. This is a very weak class of problem IMO, and does not justify the additional costs or machinery to me.

If it’s guarding against literally doing the wrong thing, I just don’t see how a Future prevents a programmer error. 

Can you give me an example of the kind of erroneous completion you imagine, that a Future would prevent?

Reply:
If the issue is that you are worried about _not_ completing, or about ensuring each `return` has to have considered whether it passes responsibility down or invokes itself, we could introduce a special marker type returned by callbacks, so that the caller must either invoke the callback, invoke a method that invokes the callback (and returns its marker), or expressly indicate it has passed off responsibility.

e.g.
```
public class SafeConsumer<T>
{
    enum Safety { INVOKED, DEFERRED }
    Safety accept(T input)
}
```

Reply:
You’ve just listed 3 safety issues callbacks have that futures don’t. There are probably a few more if you re-read my older responses. None of these are exactly explosion factories, but they’re also unnecessary in light of the alternative. They require programmer and reviewer attention that futures don’t. Consensus protocols are difficult to understand and work on, and this code will likely be read and poked at by a lot of engineers over the coming years. To my mind, that justifies paying the (imo reasonable) performance tax for using safer concurrency primitives that are easier to reason about.

Reply:
I really don’t agree. Performance is also critical and these are *not* where the difficulty lies with distributed consensus.

I also outlined trivial solutions to these concerns?

A higher method doing the wrong thing is also just as possible with Future: you can start some work and just return another Future, for instance. This is identical to invoking the callback erroneously, and I don’t see why one is much more likely than another?

If more work is being done after invoking the lower method, what stops you doing the wrong thing with a Future?

I really couldn’t disagree more here. If you really think this class of problem is a serious issue, we just implement code style requirements that you null out a callback after passing it, and use the type system as suggested above to ensure it must be invoked before returning. A little bit of coding discipline with no cognitive burden, like all good engineering.

What additional class of problem are we unable to prevent here, with these accordances?

This isn’t a small thing either. Like streams, Future are prone to very easily disguising very costly behaviours, even ignoring their inherent costs. Once something is a Future there is a tendency to chain, and this is costly and also something we have to burn mental cycles addressing or considering. 

So the question is: are these really weak concerns, that we seem able to address quite easily, more important than ensuring the performance of the system? And I am emphatically answering no.

I am not opposed to costs for safety, but I see little to no additional safety here.

Reply:
Fair enough. I think we need to agree to disagree and find a compromise though. No hard feelings of course, I think our perspectives and concerns are both valid. We’re not going to see eye to eye on everything, and having a diversity of perspectives is a good thing.

I’ll look into it some and see if I can find a solution we can both be happy with. Or at least be equally displeased with :)

Reply:
So the C* integration actually needs something future like to properly de-dupe multiple calls to apply and read. This was discussed here https://github.com/apache/cassandra-accord/pull/10#discussion_r969953934. I ended up handling it by returning the future from the initial apply/read. This case does require supporting multiple listeners though.

I haven't tried wiring it up, but I think this class would support that use case, and address my concerns with minimal memory or cas operations: https://gist.github.com/bdeggleston/bf791b54946427b8a1f453910e7c318f

Reply:
Reviewing the C* integration more, we do need to use an actual future. It hooks into the future handling that cleans up completed futures in the single threaded cache without submitting a task to do so. It does use isDone in addition to needing multiple listeners, but doesn’t need blocking get.

I reworked the previous class into an actual future implementation that’s a bit lighter than the standard future here: https://gist.github.com/bdeggleston/3d083b0baf4d4b810d5c27f7c8de0357

Also, it seems like there’s a need for a class _like_ a partial future but without the blocking parts, which I’ve sketched out an interface for here: https://gist.github.com/bdeggleston/5992d921fe158cd3453c66c60cb3aee3. It provides support for adding callbacks, determining if the operation is done, and for retrieving the result/failure if it is. No thought has gone into the name, AsyncResult or something is probably better.

Our future could easily implement it for interoperability purposes, and I _think_ we could replace much of accord’s Future use with it (including the C* integration’s usage), since I don’t think there are many places where accord actually needs to block a thread to wait for a result, and we should be thinking twice in places where we do. The places that do need to wait typically don’t have multiple threads waiting on a single result, so that could even be implemented as a special listener. 

That should save more bytes on heap and cas operations overall than using callbacks in apply, and would address my concerns as well.

Reply:
Could you link me to the code you're referring to that _requires_ a `Future` in C*? 

This sort of speaks to my concerns, though: once `Future` gets its hooks in, it turns up everywhere, and it is hard to escape. I do not want to bake it into more APIs than necessary.

The paradigm of handling something _once the work is in-flight_ is inherently costlier than a callback that can be declared before work begins, as there is no concurrency to deal with in the latter case.

If we do end up using `Future` did you not see that I had implemented a lightweight future upthread that also reduces allocations? But it permits at most one listener, and expects single-threaded access by the producer and consumer (they may run concurrently though).

> That should save more bytes on heap and cas operations overall than using callbacks in apply

I think it is impossible for a `Future` approach to be lighter weight than a callback approach.

> I think we need to agree to disagree and find a compromise though

That would be ideal, but it might not be possible. Let's certainly try, but conflicting styles going through a library like this, and its integration with Cassandra, can cause problems. Like I've said before, we can translate _to_ a `Future` cheaply, but cannot save the cost once it is there, and `Future` infects APIs and spreads (as you have indicated above). So, if we have to choose, we should opt for the one that imposes its opinions least broadly. If anything, this conversation has made me much more strongly opposed to their use here than I was previously.

Reply:
Honestly, the more I think about it, the more I am simply -1 the use of `Future` where it isn't necessary. To some extent there has to be some design consistency, and by baking `Future` in here we are guaranteeing their widespread use forever more. I disagree with foisting this performance trade-off on the hot-path, even if we reduce its cost.

The efficient way to receive a result is to declare the receiver upfront, as there is no concurrency. There's no getting around this fundamental design flaw with `Future`, at least not without additional memory costs.

If we have to, we can have a project-level discussion around this. But, to the extent possible, I may simply invoke primary author privilege here.

Reply:
Can you link me to the code in C* that we think requires `Future`? I will see how difficult this is.

Reply:
Regarding correct usage, I would be happy to find some time (not immediately, as there are other pressures) to introduce linter checks for correct usage of callbacks, so that they are always either invoked or passed down to an inner method (that itself has the linter check run).

Would this satisfy your concerns for compiler assistance?

Reply:
> Could you link me to the code you're referring to that requires a Future in C*?

The method I’m talking about is here: https://github.com/bdeggleston/cassandra/blob/metadata-persistence-simulator/src/java/org/apache/cassandra/service/accord/AccordCommand.java#L636

> But, to the extent possible, I may simply invoke primary author privilege here.

It’s not possible to any extent, because you’re not the primary author. You and I are co-authors on this project. I don’t expect you to agree with me on everything, but you do need to interact with me professionally and respectfully.

Like you, I’ve been working on this project for years and have deep experience in the theory, design, and implementation of performant, concurrent databases and consensus protocols with a focus on correctness.

Regarding my concerns about raw callbacks, I’m not pushing back because it’s fun (it’s 100% the opposite), or in bad faith, or because I don’t know what I’m talking about. I’m pushing back because I genuinely feel this is a bad idea and pushing back is worth the effort. I spent years cleaning up subtle correctness bugs in C* 3.0, and the tendency of choosing a small performance improvements over safer code was a factor in many of the issues fixed. My experience there tells me that the performance improvement gained by using callbacks is not worth the risk, and reworking the api to effectively make them mandatory is a liability. This is subjective, but it is a valid, legitimate concern from a peer, co-author, and close co-worker. You and I are partners on this project. You don’t have to agree with me, but your disagreement doesn’t make my concern invalid, and it doesn’t mean you can unilaterally ignore it. I’d really like to find some common ground here, but if you’re not able to meet me in the middle, then your next step is to take it to the list.

Reply:
Dude, there is no disrespect here. I believe I am the primary author of this work. You can believe differently, or you can determine that it provides me no additional authority, but it is not disrespectful for me to claim it, I think it is supported by the facts. I also think when there is an _impasse_ with incompatible approaches to design, it is not unreasonable for a primary author to suggest their design approach should win out.

I think this is close to the only time I have been on the project where I have made such a suggestion, so it is no small thing, and I do not even suggest it lightly.

Regarding this dispute, I do not enjoy insisting on changes any more than you enjoy pushing back. But, we cannot have totally conflicting approaches to design. If you disagree, and believe I have zero authority on the matter, then we may have to escalate to some other arbiter - that is another option I floated.

I will note, that your claims of improved type safety are weaker than you think. A serious problem with `Future` is failing to handle exceptional completion - and this _is_ required by the type system for callbacks. This is a much bigger issue than any of the very minor - easily linted - concerns you have raised about callbacks. So, I think even on the issue of compiler assistance `Future` is notably _worse_ than a callback, since I have provided multiple mitigations to your concerns with callbacks.

Reply:
I would also like to note explicitly, that I of course respect you highly, and I do not know what may have given the impression otherwise.

Reply:
To summarise the technical disagreement as I see it, here are the problems raised with each approach:

**Callbacks**
[1] May not be invoked in some no-op scenario, where an immediate `Future` might be
[2] May be invoked erroneously

**Future** 
[3] May block threads erroneously
[4] May not handle exceptional completion
[5] _Inherently_ less efficient
[6] Is more opinionated - once we’re working with `Future` it is hard to escape, and it gets woven through the APIs. Contrariwise, it is zero cost to convert a callback to a `Future`

Now, I have demonstrated a callback API that eliminates [1], leaving only [2]. But I have yet to see an example of this kind of error, except perhaps the invoking-more-than-once variant, that we also have trivial mechanisms for.

However, I have _additionally_ proposed introducing linter support to address both [1] and [2], failing compilation if a callback is neither invoked directly nor passed to a child function on any branch.

Whereas [3] we have an example of _right here_, and [4] is something that I have seen often (indeed, it was an issue previously in this library).  I cannot see how they can be effectively addressed with linters, either.

Given the above, I think it is objectively the case that callbacks are more suitable in cases where they are sufficient. I would appreciate it if you could engage with my suggestions that address your concerns with callbacks, and justify why you think the concerns I raise with `Future` are _less_ impactful than those you have raised with callbacks.

Reply:
So, last night I went back and revisited a thought I had but discarded because it would be less efficient than callbacks. However, on further reflection I think the additional cost is probably acceptable.

The basic idea is to avoid most of the extra costs associated with a `Future` by delaying the submission of work. So instead of a `Future` chain we have essentially an async operation builder chain. I have a prototype here: https://github.com/belliottsmith/cassandra/commit/fabb1372e23ea233cb5dc487952e60d216e90774

You can only mutate the end of the chain (though we could add callbacks mid-chain if we wanted). The tail of the chain retains a reference to the head of the chain; when we add to the end of the chain, the old tail replaces its head pointer with the new tail pointer. When we execute, we replace the head pointer with the final callback, then invoke the head to start processing, which passes the result of execution down through the chain.

With respect to _safety_ I think this solves both of our concerns neatly, as we have all the same mechanisms for encapsulating responsibility that a `Future` does, but we also do not even start the work until some callback has been provided that _at least_ handles exceptions, and we have no `get` method to misuse (though we can always add this feature if we want; or add a way to convert to `Future`).

With respect to efficiency, the main additional cost is the encapsulating object that represents the deferred execution. 
Since there's no concurrency with consumer / producer, we can get away with minimal state, and can directly extend the implementing classes where efficiency matters (even for transformations). Importantly, we have zero concurrency overhead at any step of execution (as opposed to every step with `Future`).

We're still encouraging the use of lambdas and other allocating abstractions, but it's probably manageable, and as mentioned above we can reduce this where it matters.

_Personally_ I still prefer callbacks. But, since this makes each step of chaining relatively fixed-cost, I think the cognitive burden I experience reasoning about chains of operations with this abstraction will be much lower than with `Future`. 

WDYT? Would this be a proper compromise between our approaches?

Reply:
This is really neat. Yes, this addresses my concerns, and I agree that this would be a good compromise. I appreciate you working with me to find a solution we both find acceptable.

I have 2 minor admin/project requests requests. First, do you mind if we punt integration of this and unblock merging metadata-persistence? I think this is the last real blocker and I’d like to get it committed before I go on vacation in a few days. Second, do you mind if _I_ do the accord and C* integration? You’re free to work on the AsyncChain interface/implementation as you like, but I’d like to do the actual integration work to adjust to the change in mental model. There are some slightly intricate places this applies to the C* integration, and I’d feel better a) doing it myself, and b) getting a little practice before doing it.

Reply:
Absolutely to both. I agree it would be better to get things unblocked at this stage.

Also happy to leave the AsyncChain stuff to you, or to collaborate more directly when you're back. I expect the integration will inform some of the design choices, so probably better to develop it more when that gets under way.

Reply:
I may have to walk that back unfortunately. This stuff touches _ so much_ of what I'm now rebasing against, that I'm not sure I can reasonably leave it be.

Reply:
FWIW, temporarily converting to Future should accomplish 95% of the conversion to asyncchain, with just the future producer consumer needing any real changes, and the rest being a rename to asyncchain. If there’s something I’m. It thinking of that would complicate making that an intermediate step and you need to do some integration, that’s fine

Reply:
Tangentially, there may be something other than future support I’m missing, but if AsyncChain lives in accord at first and replaces future, we may also be able to eliminate the circular C* dependency, which would make discussions about incrementally merging a bit easier

Reply:
Yeah, I may be able to leave `AsyncChain` itself alone for now, but I'll be tweaking the control flow for processing of messages by command stores, as its very different with partial replication (with the concept of deferred execution when earlier messages haven't arrived, amongst other things). I had hoped to leave it all entirely untouched until we could collaborate, since it's all intertwined, but we can revisit it when you're back.